File: CS/WebDashboard_CustomExport/TableReport1.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using DevExpress.XtraReports.UI;
using System.Data;

namespace AspProject {
    public partial class TableReport1 : DevExpress.XtraReports.UI.XtraReport {
        int TableWidth { get { return PageWidth - Margins.Left - Margins.Right; } }
        public TableReport1(DataTable table, dynamic parameters) {
            InitializeComponent();

            for (int i = 0; i < parameters.Length; i++) {
                var label = new XRLabel() { WidthF = TableWidth, Text = string.Format("{0}: {1}", parameters[i].Name, parameters[i].Value) };
                label.LocationF = new PointF(0, label.HeightF * i);
                ReportHeader.Controls.Add(label);
                if (i == 0)
                    ReportHeader.HeightF = label.HeightF * parameters.Length;
            }


            AddContent(table);

        }
        private void AddHeaders(DataTable table) {
            XRTable headers = XRTable.CreateTable(new Rectangle(0, 0, TableWidth, 35), 1, 0);
            headers.BeginInit();
            foreach (DataColumn dc in table.Columns) {
                XRTableCell cell = new XRTableCell();
                cell.HeightF = 35f;
                cell.CanGrow = false;
                cell.Text = dc.ColumnName;
                cell.BackColor = Color.LightGray;
                headers.Rows.FirstRow.Cells.Add(cell);
            }
            headers.HeightF = 35f;
            headers.EndInit();
            ReportHeader.Controls.Add(headers);
        }

        private void AddContent(DataTable table) {

            DataSource = table;

            int colCount = table.Columns.Count;
            int colWidth = (PageWidth - (Margins.Left + Margins.Right)) / colCount;


            XRTable tableHeader = new XRTable();
            tableHeader.Width = (PageWidth - (Margins.Left + Margins.Right));
            tableHeader.HeightF = 35;
            tableHeader.LocationF = new PointF(0, ReportHeader.HeightF);
            ReportHeader.HeightF = ReportHeader.HeightF + tableHeader.HeightF;

            XRTableRow headerRow = new XRTableRow();
            tableHeader.Width = tableHeader.Width;
            tableHeader.Rows.Add(headerRow);
            tableHeader.BeginInit();
            
            // Create a table to display data
            XRTable tableDetail = new XRTable();
            tableDetail.Width = (PageWidth - (Margins.Left + Margins.Right));

            XRTableRow detailRow = new XRTableRow();
            detailRow.Width = tableDetail.Width;
            tableDetail.Rows.Add(detailRow);
            tableDetail.BeginInit();

           

            // Create table cells, fill the header cells with text, bind the cells to data
            for (int i = 0; i < colCount; i++) {
                XRTableCell headerCell = new XRTableCell();
                headerCell.Width = colWidth;
                headerCell.Text = table.Columns[i].Caption;
                headerCell.BackColor = Color.LightGray;
                headerRow.Cells.Add(headerCell);
                

                XRTableCell detailCell = new XRTableCell();

                detailCell.Width = colWidth;
                detailCell.DataBindings.Add("Text", null, table.Columns[i].Caption);
                if (i == 0) {
                    headerCell.Borders = DevExpress.XtraPrinting.BorderSide.Left | DevExpress.XtraPrinting.BorderSide.Top | DevExpress.XtraPrinting.BorderSide.Bottom;
                    detailCell.Borders = DevExpress.XtraPrinting.BorderSide.Left | DevExpress.XtraPrinting.BorderSide.Top | DevExpress.XtraPrinting.BorderSide.Bottom;
                }
                else {
                    headerCell.Borders = DevExpress.XtraPrinting.BorderSide.All;
                    detailCell.Borders = DevExpress.XtraPrinting.BorderSide.All;
                }

                // Place the cells into the corresponding tables
                detailRow.Cells.Add(detailCell);
            }

            tableHeader.EndInit();
            tableDetail.EndInit();
            ReportHeader.Controls.Add(tableHeader);
            Detail.Controls.Add(tableDetail);
            Detail.HeightF = 23; //minimal XRTableRow Height
        }
    }
}

File: VB/WebDashboard_CustomExport/TableReport1.vb
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports DevExpress.XtraReports.UI
Imports System.Data

Namespace AspProject
    Partial Public Class TableReport1
        Inherits DevExpress.XtraReports.UI.XtraReport

        Private ReadOnly Property TableWidth() As Integer
            Get
                Return PageWidth - Margins.Left - Margins.Right
            End Get
        End Property
        Public Sub New(ByVal table As DataTable, ByVal parameters As Object)
            InitializeComponent()

            For i As Integer = 0 To parameters.Length - 1
                Dim label = New XRLabel() With {.WidthF = TableWidth, .Text = String.Format("{0}: {1}", parameters(i).Name, parameters(i).Value)}
                label.LocationF = New PointF(0, label.HeightF * i)
                ReportHeader.Controls.Add(label)
                If i = 0 Then
                    ReportHeader.HeightF = label.HeightF * parameters.Length
                End If
            Next i


            AddContent(table)

        End Sub
        Private Sub AddHeaders(ByVal table As DataTable)
            Dim headers As XRTable = XRTable.CreateTable(New Rectangle(0, 0, TableWidth, 35), 1, 0)
            headers.BeginInit()
            For Each dc As DataColumn In table.Columns
                Dim cell As New XRTableCell()
                cell.HeightF = 35F
                cell.CanGrow = False
                cell.Text = dc.ColumnName
                cell.BackColor = Color.LightGray
                headers.Rows.FirstRow.Cells.Add(cell)
            Next dc
            headers.HeightF = 35F
            headers.EndInit()
            ReportHeader.Controls.Add(headers)
        End Sub

        Private Sub AddContent(ByVal table As DataTable)

            DataSource = table

            Dim colCount As Integer = table.Columns.Count
            Dim colWidth As Integer = (PageWidth - (Margins.Left + Margins.Right)) / colCount


            Dim tableHeader As New XRTable()
            tableHeader.Width = (PageWidth - (Margins.Left + Margins.Right))
            tableHeader.HeightF = 35
            tableHeader.LocationF = New PointF(0, ReportHeader.HeightF)
            ReportHeader.HeightF = ReportHeader.HeightF + tableHeader.HeightF

            Dim headerRow As New XRTableRow()
            tableHeader.Width = tableHeader.Width
            tableHeader.Rows.Add(headerRow)
            tableHeader.BeginInit()

            ' Create a table to display data
            Dim tableDetail As New XRTable()
            tableDetail.Width = (PageWidth - (Margins.Left + Margins.Right))

            Dim detailRow As New XRTableRow()
            detailRow.Width = tableDetail.Width
            tableDetail.Rows.Add(detailRow)
            tableDetail.BeginInit()



            ' Create table cells, fill the header cells with text, bind the cells to data
            For i As Integer = 0 To colCount - 1
                Dim headerCell As New XRTableCell()
                headerCell.Width = colWidth
                headerCell.Text = table.Columns(i).Caption
                headerCell.BackColor = Color.LightGray
                headerRow.Cells.Add(headerCell)


                Dim detailCell As New XRTableCell()

                detailCell.Width = colWidth
                detailCell.DataBindings.Add("Text", Nothing, table.Columns(i).Caption)
                If i = 0 Then
                    headerCell.Borders = DevExpress.XtraPrinting.BorderSide.Left Or DevExpress.XtraPrinting.BorderSide.Top Or DevExpress.XtraPrinting.BorderSide.Bottom
                    detailCell.Borders = DevExpress.XtraPrinting.BorderSide.Left Or DevExpress.XtraPrinting.BorderSide.Top Or DevExpress.XtraPrinting.BorderSide.Bottom
                Else
                    headerCell.Borders = DevExpress.XtraPrinting.BorderSide.All
                    detailCell.Borders = DevExpress.XtraPrinting.BorderSide.All
                End If

                ' Place the cells into the corresponding tables
                detailRow.Cells.Add(detailCell)
            Next i

            tableHeader.EndInit()
            tableDetail.EndInit()
            ReportHeader.Controls.Add(tableHeader)
            Detail.Controls.Add(tableDetail)
            Detail.HeightF = 23 'minimal XRTableRow Height
        End Sub
    End Class
End Namespace
